File: src/listener.cpp

#include "ros/ros.h"
#include "rate_test/SuperAwesome.h"


void chatterCallback(const rate_test::SuperAwesome::ConstPtr& msg)
{
	static ros::Time prev_time = ros::Time::now();
	static int count = 0;
	static int total_msg_count = 0;
	
	ros::Time curr_time = ros::Time::now();
	double diff_time = (curr_time - prev_time).toSec();
	count++;
	total_msg_count++;
	
	// show the rate once per second
  	if( diff_time >= 1.0)
  	{
		ROS_INFO("%d : %f / %f = %f",total_msg_count, (double)count, diff_time , count/diff_time);
		count = 0;
		prev_time = curr_time;
	}
}

int main(int argc, char **argv)
{

  ros::init(argc, argv, "listener");
  ros::NodeHandle n;
  ros::Subscriber sub = n.subscribe("chatter", 10, chatterCallback);
  ros::spin();

  return 0;
}

File: scripts/listener.py
#!/usr/bin/env python

import rospy
from std_msgs.msg import String

class Listener():
    def __init__(self):
		rospy.init_node('listener', anonymous=True)
		rospy.Subscriber('chatter', String, self.callback)
		self.count = 0
		self.total_msg_count = 0
		self.prev_time = rospy.get_rostime()

    def callback(self,data):
		now = rospy.get_rostime()
		diff_time = (now - self.prev_time).to_sec()
		self.count += 1
		self.total_msg_count += 1
		if( diff_time >= 1.0):
			rospy.loginfo("{0}: {1} / {2} = {3}".format(self.total_msg_count, self.count, diff_time, self.count/diff_time) )
			self.count = 0
			self.prev_time = now

if __name__ == '__main__':
	listener = Listener()
	rospy.spin()
